pkg/collector/corechecks/safesender.go

```diff
@@ -0,0 +1,95 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package corechecks
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+)
+
+// SafeSender implements aggregator.Sender, wrapping the methods to provide
+// some additional safety checks.
+//
+// In particular, the methods taking `tags []string` are wrapped to copy the
+// slice, as the aggregator may modify it in-place.
+type safeSender struct {
```

Comment:
Instead of a new type, should we consider modifying the original `checkSender` to embed a tags preprocessing method:

```go
type checkSender struct {
   ... all existing attributes ...
   preprocTags func([]string) []string
}
```

Where `newCheckSender` would set it to a noop while newSafeSender would set it to your implementation of `cloneTags`, and having this method used everywhere where's it's needed in the `checkSender` implementation:

```go
func (s *checkSender) Counter(metric string, value float64, hostname string, tags []string) {
	s.sendMetricSample(metric, value, hostname, s.preprocTags(tags), metrics.CounterType, false)
}
```

It's an open question (we may have to consider performance impact of an added noop func call) and not a strong opinion, just considering an alternative smaller change.

Reply:
That's a good suggestion, but separates the component that has the concern (core checks) from the component addressing the concern (aggregator), and means that anything using the aggregator must bear the costs (perf, complexity) while only one component gets the benefits.

Reply:
You're absolutely right.